core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java

```diff
@@ -235,6 +238,26 @@ private NodePartitionMap systemNodePartitionMap(
         });
     }
 
+    /**
+     * Query plans typically divide data into buckets to help split the work among Trino nodes.  How many buckets?  That is dictated by the connector, via
+     * {@link #getConnectorBucketNodeMap}.  But when that returns empty, this method should be used to determine how many buckets to create.
+     *
+     * @return The default bucket count to use when the connector doesn't provide a number.
+     */
+    private int getDefaultBucketCount(Session session, PartitioningHandle partitioningHandle)
+    {
+        // The goal is to have enough buckets to evenly distribute them across tasks created.  The simplest way to do that is to have exactly as many buckets as there are tasks;
+        // then each task gets one bucket.  So let's return the total number of tasks here.  This number, unfortunately, is hard to know precisely because of the variety of ways in
+        // which schedulers create tasks.  We therefore have to estimate.
+        if (getRetryPolicy(session) != RetryPolicy.TASK) {
+            // Pipeline schedulers typically create as many tasks as there are nodes.
```

Comment:
this is too low even for pipeline mode as it affects local scaling too

Reply:
Can you say more?  What is "local scaling"?  I see that stages schedulers create a task per stage; can this exceed the node count and is it relevant?

Reply:
I've created follow up PR https://github.com/trinodb/trino/pull/22000

Reply:
See https://github.com/trinodb/trino/blob/be29ab8ae7b712f9ba66ce169481fc4046292172/core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java#L183 for example. If `bufferCount` is larger than bucket count, then we are not fully utilizing writer concurrency within a task